File: content/6-errors/2-retry-backoff/python/main.py
import time


attempts = 0


def unstable_operation() -> str:
    global attempts
    attempts += 1
    if attempts < 3:
        raise RuntimeError("temporary failure")
    return "ok"


def main() -> None:
    max_attempts = 5
    backoff_s = 0.2

    for i in range(1, max_attempts + 1):
        try:
            result = unstable_operation()
            print("success:", result)
            return
        except RuntimeError as e:
            print(f"attempt {i} failed: {e}")
            if i == max_attempts:
                print("giving up")
                return
            time.sleep(backoff_s)
            backoff_s *= 2


main()

File: content/6-errors/2-retry-backoff/go/main.go
package main

import (
	"errors"
	"fmt"
	"time"
)

var attempts = 0

func unstableOperation() (string, error) {
	attempts++
	if attempts < 3 {
		return "", errors.New("temporary failure")
	}
	return "ok", nil
}

func main() {
	maxAttempts := 5
	backoff := 200 * time.Millisecond

	for i := 1; i <= maxAttempts; i++ {
		result, err := unstableOperation()
		if err == nil {
			fmt.Println("success:", result)
			return
		}

		fmt.Printf("attempt %d failed: %v\n", i, err)
		if i == maxAttempts {
			fmt.Println("giving up")
			return
		}

		time.Sleep(backoff)
		backoff *= 2
	}
}

